### PR TITLE
Add reserved global flags registry for extensions

### LINKED ISSUE #7307: Add reserved global flags registry for extensions
Extensions can currently reuse short flags that azd uses globally (`-e`, `-C`, `-o`, etc.) without any warning or enforcement. This caused #7271 when a fix for #7034 assumed `-e` was always an environment name.

We need a formal registry of reserved global flags so extensions know which short flags are off-limits, and so azd can detect conflicts.

### What this includes
- `ReservedFlag` registry in `internal/reserved_flags.go` listing all global short flags
- Helpers: `IsReservedShortFlag()`, `IsReservedLongFlag()`
- Documentation for extension authors listing reserved flags
- Optionally: build-time or install-time validation that warns when an extension uses a reserved flag

### Reserved flags (initial set)
| Short | Long | Purpose |
|-------|------|---------|
| `-e` | `--environment` | azd environment name |
| `-C` | `--cwd` | working directory |
| `-o` | `--output` | output format |
| `-h` | `--help` | help |
| | `--debug` | debug mode |
| | `--no-prompt` | non-interactive |
| | `--docs` | open docs |

Related: #7034, #7271, #7035, #7274

### PR DESCRIPTION
**Pull request by jongio**
## Summary

Adds a formal registry of reserved global short flags that extensions must not reuse. This prevents future conflicts like the `-e` collision that caused #7271.

### What's included

- `cli/azd/internal/reserved_flags.go` - `ReservedFlag` struct with 9 entries (`-e`, `-C`, `-o`, `--debug`, `--no-prompt`, etc.), lookup maps, and helper functions (`IsReserved`, `GetReserved`)
- `cli/azd/internal/reserved_flags_test.go` - 7 test functions with 29 subtests covering all lookup paths and edge cases

### Why

@vhvb1989 raised this concern on #7035 - there should be a way to block extensions from using known global flags. This is the foundation for that enforcement.

Closes #7307

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: cli/azd/internal/reserved_flags.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	{Long: "environment", Short: "e", Description: "The name of the environment to use."},
+	{Long: "cwd", Short: "C", Description: "Sets the current working directory."},
+	{Long: "debug", Short: "", Description: "Enables debugging and diagnostics logging."},
+	{Long: "no-prompt", Short: "", Description: "Accepts the default value instead of prompting."},
+	{Long: "output", Short: "o", Description: "The output format (json, table, none)."},
+	{Long: "help", Short: "h", Description: "Help for the current command."},
+	{Long: "docs", Short: "", Description: "Opens the documentation for the current command."},
+	{Long: "trace-log-file", Short: "", Description: "Write a diagnostics trace to a file."},
+	{Long: "trace-log-url", Short: "", Description: "Send traces to an Open Telemetry compatible endpoint."},
+}
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/internal/reserved_flags.go]: Fixed in c58fccdcf. Made the slice unexported (`reservedFlags`) with a `ReservedFlags()` getter that returns a defensive copy. Maps are derived from the same unexported source of truth.

File: cli/azd/internal/reserved_flags.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	reservedShortFlags = make(map[string]ReservedFlag, len(ReservedFlags))
+	reservedLongFlags = make(map[string]ReservedFlag, len(ReservedFlags))
+	for _, f := range ReservedFlags {
+		reservedLongFlags[f.Long] = f
+		if f.Short != "" {
+			reservedShortFlags[f.Short] = f
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/internal/reserved_flags.go]: Fixed in c58fccdcf. Replaced `init()` with func-literal var initializers for both maps, eliminating the init ordering hazard. The maps now reference the unexported `reservedFlags` slice directly.

File: cli/azd/internal/reserved_flags.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+var ReservedFlags = []ReservedFlag{
+	{Long: "environment", Short: "e", Description: "The name of the environment to use."},
+	{Long: "cwd", Short: "C", Description: "Sets the current working directory."},
+	{Long: "debug", Short: "", Description: "Enables debugging and diagnostics logging."},
+	{Long: "no-prompt", Short: "", Description: "Accepts the default value instead of prompting."},
+	{Long: "output", Short: "o", Description: "The output format (json, table, none)."},
+	{Long: "help", Short: "h", Description: "Help for the current command."},
+	{Long: "docs", Short: "", Description: "Opens the documentation for the current command."},
+	{Long: "trace-log-file", Short: "", Description: "Write a diagnostics trace to a file."},
+	{Long: "trace-log-url", Short: "", Description: "Send traces to an Open Telemetry compatible endpoint."},
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/internal/reserved_flags.go]: Fixed in c58fccdcf. Updated to "OpenTelemetry-compatible" per the standard product name spelling.

**Review by spboyer — COMMENTED** [review]
## Code Review

### 1. Universal reserved-flag check will break existing extensions
**`cli/azd/pkg/azdext/reserved_flags.go:22`** — High

The registry treats `environment`, `output`, `docs`, etc. as universally reserved for every extension. `azdext.Run()` exits on any match before command execution. This is broader than what azd actually pre-parses today (`CreateGlobalFlagSet` only pre-parses `cwd`, `debug`, `no-prompt`, `trace-log-file`, `trace-log-url`). In-tree extensions like `azure.ai.agents` (uses `-e` for `--environment`, `-o` for `--output`) and `azure.ai.finetune` (uses `-e` for `--project-endpoint`) will fail at startup.

**Fix:** Only reserve flags that are truly universal, or scope the extra reservations to roots created with `NewExtensionRootCommand()` that actually register those globals.

---

### 2. `collectConflicts` skip logic masks subcommand flag conflicts
**`cli/azd/pkg/azdext/reserved_flags.go` (`collectConflicts` function)** — Medium

The skip logic uses name-based lookup (`root.PersistentFlags().Lookup(f.Name)`) to avoid flagging the SDK's own persistent flags. But before Cobra's `mergePersistentFlags` (which happens during `Execute()`, not validation), a subcommand's `cmd.Flags().VisitAll()` only yields locally-defined flags. If a subcommand locally defines a flag with the same long name as a root persistent flag (e.g., `--output` meaning "output file" instead of "output format"), the name lookup finds root's flag and skips validation entirely. The conflict goes unreported.

**Fix:** Use pointer equality instead of name equality:
```go
if rootFlag := root.PersistentFlags().Lookup(f.Name); rootFlag != nil && rootFlag == f {
    return
}
```

File: cli/azd/docs/design/extension-flag-architecture.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+1. Walks the entire command tree
+2. For each command's flags, checks both long and short names against the reserved list
+3. Skips flags on the root command's persistent flag set (those are the SDK-provided azd-compatible flags)
+4. Returns a detailed error listing all conflicts with remediation guidance
+
+Any extension built with the azd SDK gets this check automatically — no opt-in required.
+
+### Sync Test
+
+A test (`TestReservedFlagNames_SyncWithInternal`) ensures the SDK-side and host-side reserved flag lists stay in sync. If a developer adds a new global flag to one list but not the other, the test fails.
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/docs/design/extension-flag-architecture.md]: Fixed in e31806963. Updated test name reference to TestReservedFlagsInSyncWithInternal.

File: cli/azd/pkg/azdext/reserved_flags_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	for name := range sdkSet {
+		require.True(t, internalSet[name],
+			"azdext reserved_flags.go has %q but internal.ReservedFlags() does not — add it to internal/reserved_flags.go",
+			name)
+	}
+
+	for name := range internalSet {
+		require.True(t, sdkSet[name],
+			"internal.ReservedFlags() has %q but azdext reserved_flags.go does not — add it to pkg/azdext/reserved_flags.go",
+			name)
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags_test.go]: Fixed in e31806963. Rewrote the sync test to compare both long AND short names between SDK and internal registries. If either side changes a shorthand, the test will catch it.

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+		// Those are the SDK-provided azd-compatible flags and are intentional.
+		if rootFlag := root.PersistentFlags().Lookup(f.Name); rootFlag != nil {
+			return
+		}
+
+		if c, ok := checkFlag(cmd, f); ok {
+			conflicts = append(conflicts, c)
+		}
+	})
+
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go:137]: Fixed in e31806963 with a simpler approach: added cmd.PersistentFlags().VisitAll() alongside the existing cmd.Flags().VisitAll(), with a checked map to deduplicate. The root persistent flag skip logic applies to both. Added TestValidateNoReservedFlagConflicts_PersistentFlagCollision test to verify.

File: cli/azd/docs/design/extension-flag-architecture.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+|-----------|-------|------|---------|--------|-------------|
+| `environment` | `e` | string | `$AZURE_ENV_NAME` | No | The name of the environment to use |
+| `cwd` | `C` | string | `""` | No | Sets the current working directory |
+| `debug` | — | bool | `false` | No | Enables debugging and diagnostics logging |
+| `no-prompt` | — | bool | `false` | No | Accepts default value instead of prompting |
+| `output` | `o` | string | `"default"` | No | The output format (json, table, none) |
+| `help` | `h` | — | — | No | Help for the current command (cobra built-in) |
+| `docs` | — | — | — | No | Opens documentation for the current command |
+| `trace-log-file` | — | string | `""` | Yes | Write a diagnostics trace to a file |
+| `trace-log-url` | — | string | `""` | Yes | Send traces to an OpenTelemetry endpoint |
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/docs/design/extension-flag-architecture.md]: Fixed in e31806963. Restructured the spec to accurately categorize flags by registration mechanism: Pre-parsed (CreateGlobalFlagSet), Root Persistent Flags (
oot.go), and Cobra Built-in/Custom. Each section describes the actual code path.

**Comment by jongio** [comment]
Responding to @spboyer's [review](https://github.com/Azure/azure-dev/pull/7312#pullrequestreview-4011599180):

**1. Universal reserved-flag check will break existing extensions**

All 9 flags in the reserved list are legitimate azd globals: `-e/--environment`, `-C/--cwd`, `-o/--output`, `--debug`, `--no-prompt`, `--help/-h`, `--docs`, and the trace flags. You're right that `CreateGlobalFlagSet` only pre-parses a subset today, but these are all flags that azd defines at the root level. The reserved list matches the actual global flag surface, not just what's pre-parsed.

The extensions that currently collide (`azure.ai.agents` uses `-e` for `--environment`, `azure.ai.models`/`azure.ai.finetune` use `-e` for `--project-endpoint`) are all being migrated in PR #7313. No extension will break at startup because the migration lands first.

**2. collectConflicts skip logic masks subcommand flag conflicts**

Good catch. Implemented pointer equality in d14bf4b9a — now only skips when the flag is literally the same `*pflag.Flag` object inherited from root, not just any flag with the same name.

File: cli/azd/pkg/azdext/run.go [hunk 8]
```diff
@@ -61,6 +61,13 @@ func Run(rootCmd *cobra.Command, opts ...RunOption) {
...
 
+	// Validate that extension-defined flags do not collide with azd reserved global flags.
+	// This check runs before execution so extension developers see the error immediately.
+	if err := ValidateNoReservedFlagConflicts(rootCmd); err != nil {
+		printError(err)
+		os.Exit(1)
+	}
+
 	if cfg.preExecute != nil {
 		if err := cfg.preExecute(ctx, rootCmd); err != nil {
```
Comment (wbreza) [inline cli/azd/pkg/azdext/run.go:75]: **[High] Validation errors silently lost for listen-mode extensions**

When azd launches an extension via gRPC (the standard `listen` lifecycle), stderr is wired to `io.Discard`. This block only calls `printError(err)` → `os.Exit(1)`, so the detailed conflict message (listing flag names, command paths, remediation guidance) is silently dropped. The developer sees a generic "extension failed" from the host with zero explanation.

The two other error blocks in this same function (`preExecute` and `ExecuteContext`) both use `ReportError(ctx, err)` first — which sends the error back via gRPC where it actually reaches the user — then fall back to `printError`. This block should follow the same pattern.

**Suggested Fix:**
```go
if err := ValidateNoReservedFlagConflicts(rootCmd); err != nil {
    if reportErr := ReportError(ctx, err); reportErr != nil {
        printError(err)
    }
    os.Exit(1)
}
```

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+		b.WriteString(c.String())
+		b.WriteString("\n")
+	}
+	b.WriteString("Remove or rename these flags to avoid conflicts with azd's global flags.\n")
+	b.WriteString("Reserved flags: ")
+	b.WriteString(strings.Join(ReservedFlagNames(), ", "))
+	return fmt.Errorf("%s", b.String())
+}
+
+// collectConflicts recursively collects flag conflicts from the command tree.
```
Comment (wbreza) [inline cli/azd/pkg/azdext/reserved_flags.go:102]: **[Medium] `checkFlag` short-circuits on first collision type, hiding compound conflicts**

When a flag has both a short-name AND long-name collision (e.g., `--environment -e` on a subcommand conflicts on both dimensions), `checkFlag` returns after detecting the short collision. The long-name collision is never reported.

This means a developer sees "short flag `-e` is reserved," renames the short flag, re-runs, and then gets a *second* error for the long name — an unnecessary fix-and-fail cycle.

Consider collecting both conflicts before returning, so the developer can fix everything in one pass.

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+	{Long: "help", Short: "h"},
+	{Long: "docs", Short: ""},
+	{Long: "trace-log-file", Short: ""},
+	{Long: "trace-log-url", Short: ""},
+}
+
+// reservedShorts and reservedLongs are lookup maps built at init time.
+var (
+	reservedShorts map[string]string // short -> long
+	reservedLongs  map[string]bool
```
Comment (wbreza) [inline cli/azd/pkg/azdext/reserved_flags.go]: **[Low] `init()` vs func-literal inconsistency with twin file**

The resolved review feedback about init-ordering hazards was applied to `internal/reserved_flags.go` (uses func-literal var initializers), but this file still uses an explicit `init()`. No functional bug, but it's an inconsistency between two files the spec describes as "kept in sync."

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+				Command:      cmdPath,
+				FlagName:     f.Name,
+				FlagShort:    f.Shorthand,
+				ReservedLong: reservedLong,
+				Reason:       fmt.Sprintf("short flag -%s is reserved by azd for --%s", f.Shorthand, reservedLong),
+			}, true
+		}
+	}
+
+	// Check long flag collision (only if the long name is the same but used for
```
Comment (wbreza) [inline cli/azd/pkg/azdext/reserved_flags.go:165]: **[Low] `commandPath` uses O(n²) prepend-based slice building**

`parts = append([]string{c.Name()}, parts...)` allocates a new backing array on every level. Consider `append` + `slices.Reverse` per AGENTS.md modern Go guidance. Negligible in practice (command trees are shallow), but it's an anti-pattern in a new file.

File: cli/azd/pkg/azdext/run.go (hunk 8 above)
Comment (jongio) [reply cli/azd/pkg/azdext/run.go:75]: Fixed — validation errors now report via `ReportError(ctx, err)` before `os.Exit(1)`, matching the pattern used by the other error blocks in the same function.

File: cli/azd/pkg/azdext/reserved_flags.go (hunk 9 above)
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go:102]: Fixed — `checkFlag` now returns `[]FlagConflict`, collecting both short-name and long-name conflicts before returning. Developers see all issues in one pass.

File: cli/azd/pkg/azdext/reserved_flags.go (hunk 10 above)
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go]: Fixed — switched to func-literal var initializers (`var reservedShorts = func() ...{}()`) matching the pattern in `internal/reserved_flags.go`.

File: cli/azd/pkg/azdext/reserved_flags.go (hunk 11 above)
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go:165]: Fixed — using `append` + `slices.Reverse` instead of prepend-based slice building.

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+			return
+		}
+		checked[f.Name] = struct{}{}
+
+		// Skip flags that are the SDK-provided root persistent flags (same object).
+		// Use pointer equality so that a subcommand defining its own flag with the
+		// same name as a root persistent flag is still validated.
+		if rootFlag := root.PersistentFlags().Lookup(f.Name); rootFlag != nil && rootFlag == f {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go]: Good catch. Tightened the exemption: only roots created by NewExtensionRootCommand (marked with an azd-sdk-root annotation) get the pointer-equality exemption. Plain roots that manually add root persistent flags colliding with reserved globals are now validated. Added test to verify. Fixed in 3ed8901.

**Comment by vhvb1989** [comment]
## Observation: Two extension authoring models and the enforcement gap

After [PR #6856](https://github.com/Azure/azure-dev/pull/6856) introduced `NewExtensionRootCommand()` and `azdext.Run()`, there are now **two distinct ways to build azd extensions**:

1. **SDK-Managed** — uses `NewExtensionRootCommand()` / `azdext.Run()`. Gets automatic global flag registration, OTel propagation, structured error reporting, and (with this PR) reserved flag enforcement via `ValidateNoReservedFlagConflicts()`. **Go only.**

2. **Self-Managed** — the extension wires its own CLI root and parses flags directly. Works with **any language** (Python, JS, .NET, Go). This is how 6 of 9 existing in-repo extensions are built, and what the Go scaffold template (`azd x init`) generates.

The reserved flag enforcement in this PR only applies to extensions that opt into `azdext.Run()` — which is currently the minority. Extensions that skip it (or are written in any non-Go language) get **no protection** against flag collisions. There is no equivalent framework for Python, JavaScript, or .NET.

This creates a gap that extension authors need to understand when choosing how to build their extension. I've created:

- **Issue [#7405](https://github.com/Azure/azure-dev/issues/7405)** — describes the two models and the need for documentation
- **PR [#7406](https://github.com/Azure/azure-dev/pull/7406)** — adds `extension-authoring-models.md` documenting both models with tradeoffs, decision guide, and best practices

Worth considering whether the design doc in this PR (`extension-flag-architecture.md`) should reference the authoring models doc to help readers understand which extensions the enforcement applies to.

**Comment by jongio** [comment]
Great analysis. You're right — enforcement only covers SDK-managed extensions using `azdext.Run()`, which is currently 3 of 9 in-repo extensions. The other 6 use the self-managed model and bypass it.

Since all in-repo extensions are Go, the simplest path is to have each self-managed extension add a `ValidateNoReservedFlagConflicts(rootCmd)` call in their `main.go` — one line per extension. Alternatively they can migrate to `azdext.Run()` for the full SDK lifecycle.

Added a 'Scope & Limitations' section to the design doc in this PR referencing #7405 and your authoring models doc in #7406. The migration can happen incrementally as a follow-up.

File: cli/azd/internal/repository/detect_confirm_apphost.go [hunk 13]
```diff
@@ -83,13 +84,12 @@ func (d *detectConfirmAppHost) render(ctx context.Context) error {
...
-		d.console.Message(ctx, "")
-	}
-
-	d.console.Message(ctx, "azd will generate the files necessary to host your app on Azure.")
+	d.console.Message(
+		ctx,
+		"azd will generate the files necessary to host your app on Azure using "+color.MagentaString(
+			"Azure Container Apps",
+		)+".\n",
+	)
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/internal/repository/detect_confirm_apphost.go]: This file isn't part of this PR's changes - the PR modifies 8 files related to reserved flag enforcement, none of which is detect_confirm_apphost.go. Resolving as not applicable to this PR.

File: cli/azd/pkg/azdext/reserved_flags.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+// reservedFlag describes a global flag owned by azd that extensions must not reuse
+// for a different purpose.
+type reservedFlag struct {
+	Long  string
+	Short string
+}
+
+// reservedGlobalFlags is the canonical list of global flags that extensions must not reuse.
+// Keep this in sync with internal.ReservedFlags and CreateGlobalFlagSet (auto_install.go).
```
Comment (JeffreyCA) [inline cli/azd/pkg/azdext/reserved_flags.go]: Could we add a reference comment in `auto_install.go` as well?

File: cli/azd/internal/reserved_flags.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	// Description explains the flag's purpose in azd.
+	Description string
+}
+
+// reservedFlags is the canonical list of global flags that extensions must not reuse.
+// It is derived from CreateGlobalFlagSet (auto_install.go), the root command's
+// persistent flags, and the extension SDK's built-in flag set (extension_command.go).
+//
+// Keep this list in sync whenever a new global flag is added to azd.
+var reservedFlags = []ReservedFlag{
```
Comment (weikanglim) [inline cli/azd/internal/reserved_flags.go]: I'm slightly concerned with the duplication of global flags across 3 different files. Is it worth   centralizing it today? I believe we're blocked waiting on the extension updates to occur and we may have more time to clean this up further.

File: cli/azd/internal/reserved_flags.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+	{Long: "environment", Short: "e", Description: "The name of the environment to use."},
+	{Long: "cwd", Short: "C", Description: "Sets the current working directory."},
+	{Long: "debug", Short: "", Description: "Enables debugging and diagnostics logging."},
+	{Long: "no-prompt", Short: "", Description: "Accepts the default value instead of prompting."},
+	{Long: "output", Short: "o", Description: "The output format (json, table, none)."},
+	{Long: "help", Short: "h", Description: "Help for the current command."},
+	{Long: "docs", Short: "", Description: "Opens the documentation for the current command."},
+	{Long: "trace-log-file", Short: "", Description: "Write a diagnostics trace to a file."},
+	{Long: "trace-log-url", Short: "", Description: "Send traces to an OpenTelemetry-compatible endpoint."},
```
Comment (weikanglim) [inline cli/azd/internal/reserved_flags.go:37]: These feel like internals that `azd` should handle through the SDK rather than via the CLI interface.

File: cli/azd/internal/reserved_flags.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+// It is derived from CreateGlobalFlagSet (auto_install.go), the root command's
+// persistent flags, and the extension SDK's built-in flag set (extension_command.go).
+//
+// Keep this list in sync whenever a new global flag is added to azd.
+var reservedFlags = []ReservedFlag{
+	{Long: "environment", Short: "e", Description: "The name of the environment to use."},
+	{Long: "cwd", Short: "C", Description: "Sets the current working directory."},
+	{Long: "debug", Short: "", Description: "Enables debugging and diagnostics logging."},
+	{Long: "no-prompt", Short: "", Description: "Accepts the default value instead of prompting."},
+	{Long: "output", Short: "o", Description: "The output format (json, table, none)."},
```
Comment (weikanglim) [inline cli/azd/internal/reserved_flags.go:33]: The output and environment flag are NOT global in azd. They are lit up per-command based on the intended usage. How should one reconcile the differences?

File: cli/azd/pkg/azdext/reserved_flags.go (hunk 14 above)
Comment (jongio) [reply cli/azd/pkg/azdext/reserved_flags.go]: Added a cross-reference comment to `CreateGlobalFlagSet()` in `auto_install.go` pointing to both `internal/reserved_flags.go` and `pkg/azdext/reserved_flags.go` so devs adding new global flags know to update the reserved registries too.

File: cli/azd/internal/reserved_flags.go (hunk 16 above)
Comment (jongio) [reply cli/azd/internal/reserved_flags.go:37]: These are already registered as CLI flags - +""+--docs+""+ is public (visible in +""+--help+""+ on every command via +""+cobra_builder.go+""+), and the trace flags are hidden but pre-parsed by +""+CreateGlobalFlagSet+""+ in +""+uto_install.go+""+. The SDK also mirrors them in +""+xtension_command.go+""+ for host-to-extension propagation.

They need to be reserved because azd pre-parses them before dispatching to the extension. If an extension defines +""+--trace-log-file+""+ for a different purpose, azd's pre-parsing would consume the value, causing silent breakage. The reservation prevents that shadowing.

File: cli/azd/internal/reserved_flags.go (hunk 17 above)
Comment (jongio) [reply cli/azd/internal/reserved_flags.go:33]: You're right that `--environment/-e` and `--output/-o` aren't registered globally via `CreateGlobalFlagSet` - they're added per-command. But they still need to be reserved for extensions because:

1. **Short flags are the scarce resource.** If an extension uses `-e` for `--endpoint`, any azd command that adds `--environment` on the same command path creates a shorthand collision. Cobra panics on duplicate shorthands, so the extension breaks when azd evolves.
2. **User expectations.** Users expect `-e` to mean environment across azd. An extension that uses `-e` for something else creates confusion even if there's no technical collision today.

The reserved list is better understood as "namespace reservations" rather than "currently global flags." The per-command vs. global distinction matters for how azd registers them internally but not for whether extensions should avoid them.

File: cli/azd/internal/reserved_flags.go (hunk 15 above)
Comment (jongio) [reply cli/azd/internal/reserved_flags.go]: Centralized it. `internal/reserved_flags.go` is now the single source of truth. `pkg/azdext/reserved_flags.go` no longer maintains its own copy - it imports `internal.ReservedFlags()` and builds its lookup indexes from that. Removed the sync test since there's only one list now. Also added a cross-reference comment in `CreateGlobalFlagSet()` so developers know to update the registry when adding new global flags.
